File: go/luhn/luhn.go
package luhn

import (
	"strconv"
	"strings"
)

// Valid checks whether an input string is validated by Luhn's algorithm
func Valid(input string) bool {

	// Remove whitespace
	input = strings.ReplaceAll(input, " ", "")

	// Check length
	if len(input) < 2 {
		return false
	}

	var (
		even = len(input)%2 == 0
		sum  int
	)

	for _, v := range input {
		// Convert string value to int.
		// If the value is not a number, return false.
		digit, err := strconv.Atoi(string(v))
		if err != nil {
			return false
		}

		// Determine if we need to double our digit according to even bool
		if even {
			digit *= 2
			if digit > 9 {
				digit -= 9
			}
		}
		even = !even
		sum += digit
	}

	return sum%10 == 0
}

File: python/luhn/luhn.py
class Luhn:
    def __init__(self, card_num: str):
        self.nums = card_num.replace(" ", "")

    def valid(self) -> bool:
        if not self.nums.isdigit() or len(self.nums) < 2:
            return False

        even = len(self.nums) % 2 == 0
        result = 0
        for char in self.nums:
            digit = int(char)
            if even:
                digit *= 2
                if digit > 9:
                    digit -= 9
            even = not even
            result += digit

        return result % 10 == 0
